eclipse-theia/theia
# Basic playwright electron support

#### What it does

Allows to write tests against an Electron app by starting the Electron app first and then attaching Playwright to it. Also, it solves the workspace problem by creating it upfront and pointing to it on app start.

Because playwright cannot access native UI elements, we introduce an optional environment variable `THEIA_ELECTRON_DISABLE_NATIVE_ELEMENTS` that, when set to `1`, enforces the rendering of HTML-based menus and file chooser dialogs, which can be accessed by playwright.

Playwright also requires us to disable early rendering of the window, which we also do via an environment variable (`THEIA_ELECTRON_NO_EARLY_WINDOW`).

A unified TheiaAppLoader is introduced that can be used in both browser and electron tests.
All existing tests are adjusted to use the new AppLoader and can be run in Electron if the environment variable `USE_ELECTRON` is set to `true`.
For convenience, a yarn target `ui-tests-electron` has been added to the playwright `package.json`.
Furthermore, the target `ui-tests-ci` has been added (runs all playwright tests in the browser and in Electron) and the Github workflow has been adjusted.

Contributed on behalf of STMicroelectronics

#### How to test

1. Build the electron example as usual
  `yarn electron build`
2. Verify that electron is working
  `yarn electron start`
3. Run the electron playwright tests:
   ```
   cd examples/playwright
   yarn build
   yarn ui-tests-electron
   ```
   The testsuite automatically sets all required environment variables, so there is no need to do so manually.

Note that, #12405 still prevents running the playwright tests on Windows or MacOS, which means testing this PR currently requires Linux.

I noticed some flakiness in both browser and electron playwright tests (especially the ones for text-editor, terminal, and outline-view), but I could reproduce these issues on master, so in my opinion, they are not specific to changes introduced in this PR.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
